Session event log sketch

0xBEEFCAF3 opened the pull request:
This PR presents a rough sketch of the session event log architecture that @nothingmuch and I have been working on. Its primary purpose is to provide visibility into the direction we're taking with this design. I’d recommend against nitpicks or formatting comments at this stage. Architectural feedback is the most valuable form of review.

Note: **This PR is not intended to be merged**. It will be closed, and a follow-up PR will be opened with a clean commit history. The current branch history has diverged beyond repair, so the plan is to migrate functional chunks of code into a new branch. At present, this branch contains a working implementation of the session event log with replay support for the v2 receiver. We're currently focused on adding UniFFI bindings.

DanGould commented:
From my read there are a few major architectural changes to how both sender and receiver persist events, which affects how both `Sender` and `Receiver` proceed through their operation. Here's my understanding of what changed:

## Serialize all the things

Many, many structs become serializable so that they may be persisted.
 
## Idiomatic, generic `Receiver<State, Persister>` typestate pattern

`Receiver` becomes generic over its state and session persister implementation. This lets each state share a common  `process_event` function so that events can be replayed in a loop from persistence `logs` to recover the latest state.

## `ReceiverState` enum introduced

Each `ReceiverState` enum variant encapsulates an instance of `impl Receiver<State, P>` for some concrete `State`, because rust can't pattern match on generic type parameters.

## `ReceiverSessionEvent` enum introduced

Each `ReceiverSessionEvent` enum variant encapsulates serialized data necessary to make a state transition. The encapsulation allows that common `ReceiverState::process_event(event: ReceiverSessionEvent, persister)`  trait function to apply the next state in a loop that reads from the logs.

## `Token` concept removed

This change allows persistence to happen inside the state transitions without an explicit `save`/`load` that we were using before.

Instead, each session itself has a persister (`PersistedSession` as of writing) implementation itself with arbitrary load functionaly. This may address some of the confusion around how the reference implementation was written vs the needs of persisting session data as in @spacebear21's #689.

## PersistableError introduced

Somehow, some errors get recorded to show what something failed by calling `PersistedSession::record_error`. This part seems the least fleshed out and I don't have much insight into design decisions yet. I'm a bit confused as to why an error wouldn't also be an Event. Is there a good reason for that or is this part still underdesigned?

---

Please correct any misunderstanding of my read so far

0xBEEFCAF3 commented:
last push includes a sender type state pattern similar to the receiver type state design.
The possible type states include: 
```rust
pub enum SenderState<P> {
    Uninitialized(),
    WithReplyKey(Sender<SenderWithReplyKey, P>),
    GetContext(Sender<V2GetContext, P>),
    ProposalReceived(Sender<ProposalReceived, P>),
}
```

Application devs will still interact with the senderbuilder to construct a Sender with reply key however when replaying the logs internally we use a uninit sender and not the sender builder. 
One notable change which I would like concept ack/nack feedback on is I coalesced the `process_res` and `extract_req` type states on the v2 sender. Meaning when the sender is in state `WithReplyKey` they can extract req but that does not move the sender to the next state. And it does not persist anything new b/c we didnt learn anything new. Sender does get a POST context but I believe that is deterministic. Only after `process_res`'ing does the sender persist the orignal proposal to dir, persist the session event and move onto. This mimicks the same pattern we have in receiver<WithContext>. 

Additionally I added a type state `ProposalReceived` which is only attainable when we `process_res` the payjoin proposal. It marks the success state for the sender and wraps around the payjoin PSBT.

DanGould reviewed (commented):
> One notable change which I would like concept ack/nack feedback on is I coalesced the process_res and extract_req type states on the v2 sender. Meaning when the sender is in state WithReplyKey they can extract req but that does not move the sender to the next state.

I.e. sender state was created with URI, original PSBT, and reply key.

> And [`Sender<SenderWithReplyKey, P>::extract_*`] does not persist anything new b/c [it] didnt learn anything [worth persisting].

Although `ohttp_relay` is passed to `Sender<...>::extract_v2`, it doesn't need to be persisted. Got it. *Assuming* new persistence is the only reason to advance state, this function does not advance state. Makes sense.

> [`Sender::extract_v2] does [return] a [V2PostContext] but I believe that is deterministic.

I don't believe it it is deterministic because  `encrypt_message_a` calls `hpke::setup_sender`, where it calls `Kem::encap`, which depends on csprng. Similarly, OHTTP encapsulation is not deterministic otherwise 2 requests encapsulating the same body would be linkable.

> Only after `process_res`'ing does the sender persist the orignal proposal to dir, persist the session event and move onto.

So the original PSBT is not persisted after the builder builds? Or the proposal PSBT is only persisted after it's received?

> This mimicks the same pattern we have in receiver.

Is mimicking the pattern the rationale for the change? The rationale for state transition upon v0.23.0 `Sender::extract_req` was that only 1 request would ever need to be extracted, and there was indeed a new state that needed to wait for a response. The sender's request would not poll. However, if the request failed, or if there were no network access before process_res could be called, I could see wanting to be able to extract another request instead of failing the session or persisting that request for replay (which would break unlinkability between OHTTP requests)

Is there some other rationale relating to persistence being a requirement for state transition proposed here? I'm not sure I have enough information to cack/nack, since I'm not sure exactly what you mean by "I coalesced the process_res and extract_req type states" or what the rationale for doing so is.

0xBEEFCAF3 commented:
The latest push reintroduces storage tokens, now referred to as *state transition objects*. The key change is that instead of persisting session events directly within the type state during a transition function (e.g., `process_res`), the transition function returns a state transition object. This object is then evaluated by the persister, which handles persisting the associated session event and returning the next state.

This approach introduces some additional boilerplate for application developers but enforces a clean separation between state logic and I/O operations by moving persistence concerns outside the type state.

State transition objects are grouped into six categories. Naming is provisional and we can bike shed later.

```rust
// A transition that can be either fatal or transient or have no results.
/// If success it must have a next state and an event to save
pub type MaybeFatalTransitionWithNoResults<Event, NextState, CurrentState, Err> =
    MaybeFatalStateTransitionResultWithNoResults<Event, NextState, CurrentState, Err>;

/// A transition that can be either fatal or transient.
/// If success it must have a next state and an event to save
/// If Fatal error we must save the event and close the session
pub type MaybeFatalTransition<Event, NextState, Err> =
    MaybeFatalStateTransitionResult<Event, NextState, Err>;

/// A transition that can be transient.
/// If success it must have a next state and an event to save
pub type MaybeTransientTransition<Event, NextState, Err> =
    Result<AcceptNextState<Event, NextState>, RejectTransient<Err>>;

/// A transition that can be success or transient error
/// If success there are no events to save or next state
/// If transient error we can retry this state transition
pub type MaybeSuccessTransition<Err> = Result<AcceptCompleted, RejectTransient<Err>>;

/// A transition that is always a next state transition
pub type NextStateTransition<Event, NextState> = AcceptNextState<Event, NextState>;

/// A transition that can be success or bad init inputs
/// This is a special case where the state machine init inputs are bad and we can't proceed
/// The only thing we can do is reject the session. Since the session doesnt really exist at this point
/// there is no need to save events or close the session
pub type MaybeBadInitInputsTransition<Event, NextState, Err> =
    Result<AcceptNextState<Event, NextState>, RejectBadInitInputs<Err>>;
```

Each transition type captures a specific set of possible outcomes that a state transition function may return. These include:

1. The session is completed and closed.
2. The session progresses to the next state, and the event must be saved before the new type state is obtained.
3. A transient error occurs; the session is not closed and the state does not change. A retry is expected.
4. A fatal error occurs; the session is closed and the event must be persisted.
5. The state machine initialization fails due to invalid input. The session is rejected without saving any event or transitioning state.

From the application developer’s perspective, the only implementation requirement remains the `save_event` function on the persister. Each persister function is now statically typed to its associated transition object, abstracting the underlying details of how each transition is persisted.

A notable detail is that transition types which include fatal outcomes are not wrapped in `Result`. This is a deliberate choice to prevent misuse such as blindly propagating (`?`) or unwrapping transitions that require events to be saved. 

The remaining work before this PR can be broken into smaller, reviewable units includes:

* Refactoring all state transition functions to consume `self`.
* Applying the same state transition abstraction model to the sender.
* Resolving potential asymmetries between the sender and receiver state machines.

0xBEEFCAF3 commented:
One hazy idea: in general its a tough to know what persister method you need to call for what state transition type alias. Two solutions
* Either all state transition methods are structs and the persister save methods are named accordingly. E.g create_session give you a MaybeBadInitInputs struct which wraps a Result type and the persister has a method save_maybe_bad_init_inputs
* Or each state transition method returns a struct with save() method which consumes a ref persister which then give you your next type state 

cc @nothingmuch

0xBEEFCAF3 commented:
The latest push implement approach two [here](https://github.com/payjoin/rust-payjoin/pull/675#issuecomment-2884265975)

Much happier with the reduction in boiler plate. Example: 

```rust
let proposal = proposal
            .check_broadcast_suitability(None, |tx| Ok(wallet.can_broadcast(tx)?))
            .save(persister);
```

0xBEEFCAF3 commented:
* Rebased on master (as of this morning. I missed a couple commits). 
* Fixed the int. tests. 
* Split the persister session trait into two. A publicly accessible trait with methods to implement (save_event, load, and close) and an internal trait that extends the public one. The internal trait pub(crate)'s the methods that handle specific state transition objects. 
* Reduces api side boilerplace when returning state transition objects by proving util methods for expressing the specific state transition.
e.g  
```diff
- return MaybeFatalStateTransitionResult::Err(
-     MaybeFatalRejection::Transient(RejectTransient(
-           Error::ReplyToSender(e),
-       )),
-     );
+ return ReplyableError::Implementation(_) => MaybeFatalTransition::transient(Error::ReplyToSender(e))
```

0xBEEFCAF3 commented:
This PR has served its purpose. Most changes have been applied to master. 
Related tracking item: #785
